### DISCUSSION TITLE
Support custom APIRouter subclass

### DISCUSSION THREAD

**Discussion by christhekeele** [fastapi/fastapi#8840] [Questions]
### Description

One feature many web frameworks provide is a way to have all `GET` routes automatically define a `HEAD` version.

I'd like to modify my application's route registration to support this in a way I don't *think* I can accomplish via [middleware](https://fastapi.tiangolo.com/advanced/middleware/), [request, or route](https://fastapi.tiangolo.com/advanced/custom-request-and-route/) customization to support this. An application's `APIRouter` instance (`self.router`) is what's responsible for this, and the main application's [top-level router class](https://github.com/tiangolo/fastapi/blob/22528373bba6a654323de416ad5c867cbadb81bb/fastapi/applications.py#L69) is hardcoded.

### The solution you would like

It'd be nice if I could parameterize a FastAPI app's top-level router with a custom `Router` subclass:

```Python
import fastapi

class Router(fastapi.routing.APIRouter):

  @staticmethod
  def generate_head_from_get(route): #...

  def add_api_route(self, methods, **kwargs):
    if "GET" in methods:
      # super and add derived HEAD implementation
    else:
       # super
     
    
app = fastapi.FastAPI(router_class=router)


@app.get("/", teleport_to_moon=True)
def read_root():
    return {"Hello": "World"}
```

Then I could `curl -IHEAD localhost:8000` without further effort.

### Describe alternatives you've considered

One way to do this now today is to overwrite the [`app.add_api_route`](https://github.com/tiangolo/fastapi/blob/22528373bba6a654323de416ad5c867cbadb81bb/fastapi/applications.py#L231) function to use my custom subclass, or subclass FastAPI itself and provide my subclass implementation of it. This is obviously terrible.

An alternative is to simply never directly add routes to your application, but always define an intermediary router subclass instance and then mount it:

```py
app = FastAPI()
router = CustomRouter()
app.include_router(router)
```

That works for my use-case since I am using manually instantiated routers everywhere! But I can see others not wanting to do that for simple apps.

### Environment

* OS: macOS
* FastAPI Version: 0.63.0
* Python version: 3.9.4

### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

**Comment by christhekeele** [comment]
I think this'd be as simple as [this](https://github.com/tiangolo/fastapi/pull/3080/files#diff-2acad16c8816cd891697cf09bc6158a64aeab4091ad4cb7a3e5cce88bb9a1a78)

**Comment by christhekeele** [comment]
Indeed it was. If welcome, this enhancement is available in #3080.

**Comment by YuriiMotov** [comment]
https://github.com/fastapi/fastapi/pull/3080#issuecomment-1302624269
